### PR TITLE
Implemented pause/resume on sorted Lucene doc collector

### PR DESCRIPTION
**Pull request by bogensberger**

### CODE REVIEW

File: sql/src/main/java/io/crate/operation/collect/LuceneDocCollector.java [hunk 1]
```diff
@@ -207,97 +200,181 @@ public void doCollect() {
         visitorEnabled = fieldsVisitor.required();
         SearchContext.setCurrent(searchContext);
         searchContext.searcher().inStage(ContextIndexSearcher.Stage.MAIN_QUERY);
-        Query query = searchContext.query();
+        doCollect(null);
+    }
 
+    private void doCollect(OrderedCollectContext collectContext) {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/operation/collect/LuceneDocCollector.java]: annotate as nullable

File: sql/src/main/java/io/crate/operation/collect/LuceneDocCollector.java [hunk 2]
```diff
@@ -207,97 +200,181 @@ public void doCollect() {
...
-            searchContext.clearReleasables(SearchContext.Lifetime.PHASE);
-            SearchContext.removeCurrent();
         }
     }
 
     @Override
     public void kill() {
         killed = true;
+        if (paused.get()) {
+            doCollect(collectContext);
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/operation/collect/LuceneDocCollector.java]: couldn't you call `downstream.fail(new CancellationException)` directly here? starting doCollect is kinda confusing

File: sql/src/main/java/io/crate/operation/collect/LuceneDocCollector.java [hunk 3]
```diff
@@ -207,97 +200,181 @@ public void doCollect() {
...
+        if (orderBy == null) {
+            throw new UnsupportedOperationException();
+        }
+        if (paused.compareAndSet(true, false)) {
+            doCollect(collectContext);
+        }
+    }
+
+
+    private class OrderedCollectContext {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/operation/collect/LuceneDocCollector.java]: could this inner class be made static ?

File: sql/src/main/java/io/crate/operation/collect/LuceneDocCollector.java (hunk 2 above)
Comment (bogensberger) [inline sql/src/main/java/io/crate/operation/collect/LuceneDocCollector.java]: problem is that i want the catch()  { .. } finally {...} blocks from doCollect() to be called

**Comment by mfussenegger** [comment]
ack

**Pull request merged**